dbt/models/location/location.crosswalk_year_fill.sql

```diff
@@ -6,271 +6,234 @@ equivalent location data are filled thus:
+      `year` will have a null value for `chicago_ward_data_year`
+
+Taken together, these two assumptions mean that this query _only_ performs
+backfilling for all input data, and leaves forward-filling to the queries
+that produce the input tables. This might seem like a strange choice, but it
+stems from the fact that we need to forward-fill different data sources
+slightly differently, while we perform backfilling in the same way across
+all data sources.
+
+The output of this query contains three different variations on "year" columns:
+
+1. `year` is the PIN year; there should be one `year` for every year that
+   exists in `spatial.parcel`
+2. `{feature_name}_fill_year` is the year of data for a PIN in the input data
+   that should be used to fill records with a given `year`
+3. `{feature_name}_data_year` is the original year in the raw data that
+   produced the value that exists in the fill year
+
+Take this hypothetical row as an example:
+
+| year | census_fill_year | census_data_year |
+| ---- | ---------------- | ---------------- |
+| 2019 | 2015             | 2017             |
+
+This row means that for a PIN in year 2010, a data consumer can get a filled
+value for a `location.census` feature by joining to the representation of that
+PIN in the `location.census` table with year 2015, and that this particular row
+of data in `location.census` uses a value that was recorded in the raw data in
+2017.
+
+So if we want to use the crosswalk to fill data in this case, we can write a
+query like this:
+
+    SELECT *
+    FROM spatial.parcel AS pin
+    INNER JOIN location.crosswalk_year_fill AS cyf
+        ON pin.year = cyf.year
+    LEFT JOIN location.census AS census
+        ON pin.pin10 = census.pin10
+        AND cyf.census_fill_year = census.year
 */
 {{ config(materialized='table') }}
 
+{#-
+    Mapping containing of the fields we want to fill, where each key is the
+    name of a table in the `location` schema and each value is a list of names
+    of `*_data_year` fields in that table that we want to use for filling
+-#}
+{%- set fields = {
+    "census": ["census"],
+    "census_acs5": ["census_acs5"],
+    "political": [
+        "cook_board_of_review_district",
+        "cook_commissioner_district",
+        "cook_judicial_district",
+        "ward_chicago",
+        "ward_evanston"
+    ],
+    "chicago": [
+        "chicago_community_area",
+        "chicago_industrial_corridor",
+        "chicago_police_district"
+    ],
+    "economy": [
+        "econ_coordinated_care_area",
+        "econ_enterprise_zone",
+        "econ_industrial_growth_zone",
+        "econ_qualified_opportunity_zone",
+        "econ_central_business_district"
+    ],
+    "environment": [
+        "env_flood_fema",
+        "env_flood_fs",
+        "env_ohare_noise_contour",
+        "env_airport_noise"
+    ],
+    "school": ["school"],
+    "tax": ["tax"],
+    "access": ["access_cmap_walk"],
+    "other": ["misc_subdivision"]
+} -%}
```

Comment:
I noticed while editing this model that it contains a ton of boilerplate code, so I factored out all of the duplicated code using Jinja loops and this central data structure. Hopefully I haven't traded off clarity for brevity; if you think I have, I'm happy to revert back to the duplicative version of the code.